File: android/app/src/main/java/com/signal/app/ui/screens/chat/components/QuickActionPanel.kt
package com.signal.app.ui.screens.chat.components

import androidx.compose.animation.*
import androidx.compose.animation.core.*
import androidx.compose.foundation.*
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.scale
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.hapticfeedback.HapticFeedbackType
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.unit.dp
import com.signal.app.data.model.chat.QuickReplyType
import com.signal.app.ui.theme.*

@Composable
fun QuickActionPanel(
    isVisible: Boolean,
    onQuickReply: ((QuickReplyType) -> Unit)? = null,
    onLocationShare: (() -> Unit)? = null,
    onImagePicker: (() -> Unit)? = null,
    modifier: Modifier = Modifier
) {
    val haptic = LocalHapticFeedback.current

    AnimatedVisibility(
        visible = isVisible,
        enter = slideInVertically(
            initialOffsetY = { it },
            animationSpec = tween(
                durationMillis = 300,
                easing = LinearOutSlowInEasing
            )
        ) + fadeIn(
            animationSpec = tween(
                durationMillis = 200,
                easing = LinearEasing
            )
        ),
        exit = slideOutVertically(
            targetOffsetY = { it },
            animationSpec = tween(
                durationMillis = 200,
                easing = FastOutLinearInEasing
            )
        ) + fadeOut(
            animationSpec = tween(
                durationMillis = 150
            )
        ),
        modifier = modifier
    ) {
        Surface(
            color = ChatColors.Surface,
            shadowElevation = ChatShadows.Popup,
            modifier = Modifier.fillMaxWidth()
        ) {
            Column(
                modifier = Modifier.padding(ChatDimensions.PaddingMD)
            ) {
                // Quick Reply Section
                QuickReplySection(
                    onQuickReply = { type ->
                        haptic.performHapticFeedback(HapticFeedbackType.TextHandleMove)
                        onQuickReply?.invoke(type)
                    }
                )
                
                HorizontalDivider(
                    modifier = Modifier.padding(vertical = ChatDimensions.PaddingSM),
                    color = ChatColors.Divider
                )
                
                // Media Actions Section
                MediaActionsSection(
                    onLocationShare = {
                        haptic.performHapticFeedback(HapticFeedbackType.TextHandleMove)
                        onLocationShare?.invoke()
                    },
                    onImagePicker = {
                        haptic.performHapticFeedback(HapticFeedbackType.TextHandleMove)
                        onImagePicker?.invoke()
                    }
                )
            }
        }
    }
}

@Composable
private fun QuickReplySection(
    onQuickReply: (QuickReplyType) -> Unit
) {
    Column {
        Text(
            text = "빠른 응답",
            style = ChatTextStyles.SystemText.copy(
                color = ChatColors.TextPrimary
            ),
            modifier = Modifier.padding(bottom = ChatDimensions.PaddingSM)
        )
        
        Column(
            verticalArrangement = Arrangement.spacedBy(ChatDimensions.PaddingSM)
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(ChatDimensions.PaddingSM)
            ) {
                QuickReplyChip(
                    text = "도착했어요",
                    icon = Icons.Default.LocationOn,
                    color = ChatColors.Online,
                    onClick = { onQuickReply(QuickReplyType.ARRIVED) },
                    modifier = Modifier.weight(1f)
                )
                
                QuickReplyChip(
                    text = "가는 중",
                    icon = Icons.Default.DirectionsRun,
                    color = ChatColors.Primary,
                    onClick = { onQuickReply(QuickReplyType.ON_WAY) },
                    modifier = Modifier.weight(1f)
                )
            }
            
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(ChatDimensions.PaddingSM)
            ) {
                QuickReplyChip(
                    text = "5분 늦음",
                    icon = Icons.Default.AccessTime,
                    color = ChatColors.Away,
                    onClick = { onQuickReply(QuickReplyType.LATE_5MIN) },
                    modifier = Modifier.weight(1f)
                )
                
                QuickReplyChip(
                    text = "15분 늦음",
                    icon = Icons.Default.AccessTime,
                    color = ChatColors.Away,
                    onClick = { onQuickReply(QuickReplyType.LATE_15MIN) },
                    modifier = Modifier.weight(1f)
                )
            }
            
            QuickReplyChip(
                text = "취소할게요",
                icon = Icons.Default.Cancel,
                color = ChatColors.Accent,
                onClick = { onQuickReply(QuickReplyType.CANCEL) },
                modifier = Modifier.fillMaxWidth()
            )
        }
    }
}

@Composable
private fun QuickReplyChip(
    text: String,
    icon: ImageVector,
    color: Color,
    onClick: () -> Unit,
    modifier: Modifier = Modifier
) {
    var isPressed by remember { mutableStateOf(false) }
    val scale by animateFloatAsState(
        targetValue = if (isPressed) 0.95f else 1f,
        animationSpec = tween(150),
        label = "chip_scale"
    )

    Surface(
        onClick = onClick,
        modifier = modifier.scale(scale),
        color = color.copy(alpha = 0.1f),
        shape = RoundedCornerShape(ChatDimensions.RadiusXL),
        border = BorderStroke(
            width = 1.dp,
            color = color.copy(alpha = 0.3f)
        )
    ) {
        Row(
            modifier = Modifier
                .padding(
                    horizontal = ChatDimensions.PaddingMD,
                    vertical = ChatDimensions.PaddingSM
                )
                .clickable(
                    interactionSource = remember { MutableInteractionSource() },
                    indication = null
                ) {
                    onClick()
                },
            verticalAlignment = Alignment.CenterVertically
        ) {
            Icon(
                imageVector = icon,
                contentDescription = null,
                tint = color,
                modifier = Modifier.size(16.dp)
            )
            
            Spacer(modifier = Modifier.width(ChatDimensions.PaddingSM))
            
            Text(
                text = text,
                style = ChatTextStyles.SystemText.copy(
                    color = color
                )
            )
        }
    }
}

@Composable
private fun MediaActionsSection(
    onLocationShare: () -> Unit,
    onImagePicker: () -> Unit
) {
    Column {
        Text(
            text = "미디어 및 위치",
            style = ChatTextStyles.SystemText.copy(
                color = ChatColors.TextPrimary
            ),
            modifier = Modifier.padding(bottom = ChatDimensions.PaddingSM)
        )
        
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(ChatDimensions.PaddingMD)
        ) {
            MediaActionButton(
                label = "위치 공유",
                icon = Icons.Default.LocationOn,
                color = ChatColors.LocationBackground,
                onClick = onLocationShare,
                modifier = Modifier.weight(1f)
            )
            
            MediaActionButton(
                label = "사진",
                icon = Icons.Default.PhotoCamera,
                color = ChatColors.Primary,
                onClick = onImagePicker,
                modifier = Modifier.weight(1f)
            )
        }
    }
}

@Composable
private fun MediaActionButton(
    label: String,
    icon: ImageVector,
    color: Color,
    onClick: () -> Unit,
    modifier: Modifier = Modifier
) {
    var isPressed by remember { mutableStateOf(false) }
    val scale by animateFloatAsState(
        targetValue = if (isPressed) 0.95f else 1f,
        animationSpec = tween(150),
        label = "media_button_scale"
    )

    Surface(
        onClick = onClick,
        modifier = modifier.scale(scale),
        color = color.copy(alpha = 0.1f),
        shape = RoundedCornerShape(ChatDimensions.RadiusMD),
        border = BorderStroke(
            width = 1.dp,
            color = color.copy(alpha = 0.3f)
        )
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(ChatDimensions.PaddingMD)
                .clickable(
                    interactionSource = remember { MutableInteractionSource() },
                    indication = null
                ) {
                    onClick()
                },
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Box(
                modifier = Modifier
                    .size(ChatDimensions.QuickActionSize)
                    .background(
                        color = color,
                        shape = CircleShape
                    )
                    .shadow(
                        elevation = ChatShadows.Card,
                        shape = CircleShape,
                        spotColor = ChatShadows.SoftColor
                    ),
                contentAlignment = Alignment.Center
            ) {
                Icon(
                    imageVector = icon,
                    contentDescription = null,
                    tint = Color.White,
                    modifier = Modifier.size(24.dp)
                )
            }
            
            Spacer(modifier = Modifier.height(ChatDimensions.PaddingSM))
            
            Text(
                text = label,
                style = ChatTextStyles.SystemText.copy(
                    color = color
                )
            )
        }
    }
}

File: ios/lib/features/chat/presentation/widgets/quick_action_panel.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'chat_theme.dart';
import '../pages/chat_room_page.dart';

class QuickActionPanel extends StatefulWidget {
  final VoidCallback? onLocationShare;
  final VoidCallback? onImagePicker;
  final Function(QuickReplyType)? onQuickReply;
  final bool isVisible;

  const QuickActionPanel({
    Key? key,
    this.onLocationShare,
    this.onImagePicker,
    this.onQuickReply,
    this.isVisible = false,
  }) : super(key: key);

  @override
  State<QuickActionPanel> createState() => _QuickActionPanelState();
}

class _QuickActionPanelState extends State<QuickActionPanel>
    with TickerProviderStateMixin {
  late AnimationController _slideController;
  late AnimationController _fadeController;
  late Animation<double> _slideAnimation;
  late Animation<double> _fadeAnimation;

  @override
  void initState() {
    super.initState();
    _setupAnimations();
  }

  void _setupAnimations() {
    _slideController = AnimationController(
      duration: ChatDimensions.animationMedium,
      vsync: this,
    );
    
    _fadeController = AnimationController(
      duration: ChatDimensions.animationFast,
      vsync: this,
    );
    
    _slideAnimation = Tween<double>(
      begin: 1.0,
      end: 0.0,
    ).animate(CurvedAnimation(
      parent: _slideController,
      curve: Curves.easeOutBack,
    ));
    
    _fadeAnimation = Tween<double>(
      begin: 0.0,
      end: 1.0,
    ).animate(CurvedAnimation(
      parent: _fadeController,
      curve: Curves.easeIn,
    ));
  }

  @override
  void didUpdateWidget(QuickActionPanel oldWidget) {
    super.didUpdateWidget(oldWidget);
    
    if (widget.isVisible != oldWidget.isVisible) {
      if (widget.isVisible) {
        _slideController.forward();
        _fadeController.forward();
      } else {
        _slideController.reverse();
        _fadeController.reverse();
      }
    }
  }

  @override
  void dispose() {
    _slideController.dispose();
    _fadeController.dispose();
    super.dispose();
  }

  void _handleActionTap(VoidCallback? callback) {
    HapticFeedback.selectionClick();
    callback?.call();
  }

  @override
  Widget build(BuildContext context) {
    if (!widget.isVisible) {
      return const SizedBox.shrink();
    }
    
    return FadeTransition(
      opacity: _fadeAnimation,
      child: Container(
        padding: const EdgeInsets.all(ChatDimensions.paddingMD),
        decoration: const BoxDecoration(
          color: ChatColors.surface,
          boxShadow: ChatShadows.popup,
        ),
        child: SlideTransition(
          position: Tween<Offset>(
            begin: const Offset(0, 1),
            end: Offset.zero,
          ).animate(_slideAnimation),
          child: Column(
            mainAxisSize: MainAxisSize.min,
            children: [
              // Quick Reply Actions
              _buildQuickReplySection(),
              
              const Padding(
                padding: EdgeInsets.symmetric(vertical: ChatDimensions.paddingSM),
                child: Divider(color: ChatColors.divider),
              ),
              
              // Media Actions
              _buildMediaActionsSection(),
            ],
          ),
        ),
      ),
    );
  }

  Widget _buildQuickReplySection() {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Padding(
          padding: const EdgeInsets.only(bottom: ChatDimensions.paddingSM),
          child: Text(
            '빠른 응답',
            style: ChatTextStyles.systemText.copyWith(
              fontWeight: FontWeight.w600,
              color: ChatColors.textPrimary,
            ),
          ),
        ),
        
        Wrap(
          spacing: ChatDimensions.paddingSM,
          runSpacing: ChatDimensions.paddingSM,
          children: [
            _buildQuickReplyChip(
              QuickReplyType.arrived,
              '도착했어요',
              Icons.location_on,
              ChatColors.online,
            ),
            _buildQuickReplyChip(
              QuickReplyType.onWay,
              '가는 중이에요',
              Icons.directions_run,
              ChatColors.primary,
            ),
            _buildQuickReplyChip(
              QuickReplyType.late5min,
              '5분 늦을게요',
              Icons.access_time,
              ChatColors.away,
            ),
            _buildQuickReplyChip(
              QuickReplyType.late10min,
              '10분 늦을게요',
              Icons.access_time,
              ChatColors.away,
            ),
            _buildQuickReplyChip(
              QuickReplyType.late15min,
              '15분 늦을게요',
              Icons.access_time,
              ChatColors.away,
            ),
            _buildQuickReplyChip(
              QuickReplyType.cancel,
              '취소할게요',
              Icons.cancel,
              ChatColors.accent,
            ),
          ],
        ),
      ],
    );
  }

  Widget _buildQuickReplyChip(
    QuickReplyType type,
    String text,
    IconData icon,
    Color color,
  ) {
    return GestureDetector(
      onTap: () => _handleQuickReply(type),
      child: AnimatedScale(
        scale: 1.0,
        duration: ChatDimensions.animationFast,
        child: Container(
          padding: const EdgeInsets.symmetric(
            horizontal: ChatDimensions.paddingMD,
            vertical: ChatDimensions.paddingSM,
          ),
          decoration: BoxDecoration(
            color: color.withOpacity(0.1),
            borderRadius: BorderRadius.circular(ChatDimensions.radiusXL),
            border: Border.all(
              color: color.withOpacity(0.3),
              width: 1,
            ),
          ),
          child: Row(
            mainAxisSize: MainAxisSize.min,
            children: [
              Icon(
                icon,
                size: 16,
                color: color,
              ),
              const SizedBox(width: ChatDimensions.paddingSM),
              Text(
                text,
                style: ChatTextStyles.systemText.copyWith(
                  color: color,
                  fontWeight: FontWeight.w600,
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }

  Widget _buildMediaActionsSection() {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Padding(
          padding: const EdgeInsets.only(bottom: ChatDimensions.paddingSM),
          child: Text(
            '미디어 및 위치',
            style: ChatTextStyles.systemText.copyWith(
              fontWeight: FontWeight.w600,
              color: ChatColors.textPrimary,
            ),
          ),
        ),
        
        Row(
          children: [
            _buildMediaAction(
              '위치 공유',
              Icons.location_on,
              ChatColors.locationBackground,
              widget.onLocationShare,
            ),
            const SizedBox(width: ChatDimensions.paddingMD),
            _buildMediaAction(
              '사진',
              Icons.photo_camera,
              ChatColors.primary,
              widget.onImagePicker,
            ),
          ],
        ),
      ],
    );
  }

  Widget _buildMediaAction(
    String label,
    IconData icon,
    Color color,
    VoidCallback? onTap,
  ) {
    return Expanded(
      child: GestureDetector(
        onTap: () => _handleActionTap(onTap),
        child: Container(
          padding: const EdgeInsets.symmetric(
            vertical: ChatDimensions.paddingMD,
          ),
          decoration: BoxDecoration(
            color: color.withOpacity(0.1),
            borderRadius: BorderRadius.circular(ChatDimensions.radiusMD),
            border: Border.all(
              color: color.withOpacity(0.3),
              width: 1,
            ),
          ),
          child: Column(
            mainAxisSize: MainAxisSize.min,
            children: [
              Container(
                padding: const EdgeInsets.all(ChatDimensions.paddingMD),
                decoration: BoxDecoration(
                  color: color,
                  shape: BoxShape.circle,
                ),
                child: Icon(
                  icon,
                  color: Colors.white,
                  size: 24,
                ),
              ),
              const SizedBox(height: ChatDimensions.paddingSM),
              Text(
                label,
                style: ChatTextStyles.systemText.copyWith(
                  color: color,
                  fontWeight: FontWeight.w600,
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }

  void _handleQuickReply(QuickReplyType type) {
    HapticFeedback.selectionClick();
    widget.onQuickReply?.call(type);
  }
}

// Quick Action Floating Button
class QuickActionFloatingButton extends StatefulWidget {
  final VoidCallback? onPressed;
  final bool isExpanded;

  const QuickActionFloatingButton({
    Key? key,
    this.onPressed,
    this.isExpanded = false,
  }) : super(key: key);

  @override
  State<QuickActionFloatingButton> createState() => _QuickActionFloatingButtonState();
}

class _QuickActionFloatingButtonState extends State<QuickActionFloatingButton>
    with SingleTickerProviderStateMixin {
  late AnimationController _rotationController;
  late Animation<double> _rotationAnimation;

  @override
  void initState() {
    super.initState();
    _setupAnimations();
  }

  void _setupAnimations() {
    _rotationController = AnimationController(
      duration: ChatDimensions.animationMedium,
      vsync: this,
    );
    
    _rotationAnimation = Tween<double>(
      begin: 0.0,
      end: 0.375, // 45도 회전 (1/8 회전)
    ).animate(CurvedAnimation(
      parent: _rotationController,
      curve: Curves.easeInOut,
    ));
  }

  @override
  void didUpdateWidget(QuickActionFloatingButton oldWidget) {
    super.didUpdateWidget(oldWidget);
    
    if (widget.isExpanded != oldWidget.isExpanded) {
      if (widget.isExpanded) {
        _rotationController.forward();
      } else {
        _rotationController.reverse();
      }
    }
  }

  @override
  void dispose() {
    _rotationController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: () {
        HapticFeedback.selectionClick();
        widget.onPressed?.call();
      },
      child: AnimatedBuilder(
        animation: _rotationAnimation,
        builder: (context, child) {
          return Transform.rotate(
            angle: _rotationAnimation.value * 2 * 3.14159, // Convert to radians
            child: Container(
              width: ChatDimensions.quickActionSize,
              height: ChatDimensions.quickActionSize,
              decoration: BoxDecoration(
                gradient: ChatColors.instagramGradient,
                shape: BoxShape.circle,
                boxShadow: ChatShadows.card,
              ),
              child: Center(
                child: Icon(
                  widget.isExpanded ? Icons.close : Icons.add,
                  color: Colors.white,
                  size: ChatDimensions.quickActionIconSize,
                ),
              ),
            ),
          );
        },
      ),
    );
  }
}